### PR TITLE
[WIP][BEAM-7336] Add schema inferring for KafkaIO when reading Avro values

### PR DESCRIPTION
**Pull request by aromanenko-dev**
Adds `withAvroSchemaValues()` to read `PCollection<GenericRecord>` from Kafka topics where values are serialised by Avro schemas . Also, it sets Beam schema inferred on Avro schema used for values serialisation, so, it can be used as an input for `SqlTransform` downstream of pipeline. It has to be used with `ConfluentSchemaRegistryDeserializerProvider`.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by aromanenko-dev** [comment]
@iemejia 
CC: @reuvenlax 
PTAL

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -845,11 +853,23 @@ public void setValueDeserializer(String valueDeserializer) {
...
       return new TypedWithoutMetadata<>(this);
     }
 
+    /**
+     * Returns a {@link PTransform} for {@link PCollection} of {@link GenericRecord}, dropping Kafka
+     * keys and metadata. It has to be used only with {@link
+     * ConfluentSchemaRegistryDeserializerProvider}. See {@link KafkaIO} for more information on
+     * usage and configuration of reader.
+     */
+    public PTransform<PBegin, PCollection<GenericRecord>> withAvroSchemaValues() {
```
Comment (iemejia) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: We do not need an special PTransform for this. Notice that both SqlTransform and most Schema-based PTransforms do require PCollection<T> but with the specific condition that this PCollection should be encoded using a `SchemaCoder`. So we can just use the existing methods that produce `PCollection<K>` and `PCollection<V>`.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 2]
```diff
@@ -1043,6 +1063,54 @@ public void populateDisplayData(DisplayData.Builder builder) {
     }
   }
 
+  /**
+   * A {@link PTransform} to read values serialised with Avro schema from Kafka topics. It removes
+   * Kafka metadata and returns a {@link PCollection} of only values typed as {@link GenericRecord}.
+   * Also, it sets Beam schema based on Avro schema used for values serialisation. See {@link
+   * KafkaIO} for more information on usage and configuration of reader.
+   */
+  public static class ReadValuesWithAvroSchema<K>
```
Comment (iemejia) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Remove (see reasoning above)

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java [hunk 3]
```diff
@@ -492,6 +498,45 @@ public void testReadAvroSpecificRecordsWithConfluentSchemaRegistry() {
...
+            .withConsumerFactoryFn(
+                new ConsumerFactoryFn(
+                    ImmutableList.of(topic),
+                    1,
+                    numElements,
+                    OffsetResetStrategy.EARLIEST,
+                    i -> ByteBuffer.wrap(new byte[4]).putInt(i).array(),
+                    new ValueAvroSerializableFunction(topic, schemaRegistryUrl)))
+            .withMaxNumRecords(numElements)
+            .withAvroSchemaValues();
```
Comment (iemejia) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java]: You can replace this `.withoutMetadata()` chained with `Keys.create()` and/or `Values.create()` to extract the respective to be queried after.

**Comment by iemejia** [comment]
Some extra things:
- We should probably annotate the two infer methods in KafkaIO as `@Experimental(Kind.SCHEMAS)`
- The `getCoder(CoderRegistry)` method should now receive a `SchemaRegistry` too, you can get that one from the input like this: `SchemaRegistry schemaRegistry = input.getPipeline().getSchemaRegistry();`
- We can simplify the test, because the only thing the code should care about is that if we infer a Beam `SchemaCoder` and it is available then the returned PCollection has a Coder that is an instance of `SchemaCoder`. The fact that you can do SQL queries or Schema-based PTransforms from a `PCollection` encoded with a `SchemaCoder` is already given and it is a nice guarantee of Reuven's work, so we do not need to double test that here.

**Comment by iemejia** [comment]
Created #10974 which is an util method that could prove handy for this PR (to cover for example the case of Kafka reads of JavaBeans that have an associated Beam Schema). PTAL @aromanenko-dev

**Pull request closed without merging**